### PR TITLE
Fix #1471 Add double tap gesture

### LINKED ISSUE #1471: Add double tap gesture
There is currently no double tap gesture.

### PR DESCRIPTION
**Pull request by krisgiesing**
Also fixes some issues with single-tap rejection.

### CODE REVIEW

**Comment by krisgiesing** [comment]
Probably needs tests

File: sky/packages/sky/lib/src/gestures/double_tap.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    : super(router: router, deadline: kTapTimeout);
+
+  GestureTapListener onDoubleTap;
+  int _numTaps = 0;
+  Timer _longTimer;
+
+  void resolve(GestureDisposition disposition) {
+    super.resolve(disposition);
+    if (disposition == GestureDisposition.rejected) {
+      _numTaps = 0;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/double_tap.dart]: We probably need to cancel _longTimer here too.

**Comment by abarth** [comment]
What happens in the case where you have both an onTap and an onDoubleTap listener?  It seems like we'll never get double taps because the single tap detector will declare victory each time on pointer up.

File: sky/packages/sky/lib/src/gestures/double_tap.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  void didExceedLongDeadline() {
+    _numTaps = 0;
+    _longTimer = null;
+  }
+
+  void handlePrimaryPointer(sky.PointerEvent event) {
+    if (event.type == 'pointerup') {
+      _numTaps++;
```
Comment (abarth) [inline sky/packages/sky/lib/src/gestures/double_tap.dart]: Presumably we'll need to factor kDoubleTapSlop into here somehow.

**Comment by krisgiesing** [comment]
So should the tap detector invoke the tap callback but not declare victory?

Another approach might be to have a single tap detector per GestureRecognizer, which is configurable for the number of taps desired.  But I'm not sure that avoids the problem since you can have multiple GestureRecognizers in the tree.

File: sky/packages/sky/lib/src/gestures/double_tap.dart (hunk 2 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/gestures/double_tap.dart]: I actually wasn't sure what kDoubleTapSlop was supposed to represent.  Is that the maximum distance between the two down events for the two taps?

File: sky/packages/sky/lib/src/gestures/double_tap.dart (hunk 1 above)
Comment (krisgiesing) [inline sky/packages/sky/lib/src/gestures/double_tap.dart]: Good catch

**Comment by krisgiesing** [comment]
If we have a tap detector in place but not a double tap detector, and someone does a double tap, should two separate tap event callbacks be invoked?  How about the same scenario but you add a double tap detector?

A third approach would be to add information to the tap callback that returns the number of consecutive taps recognized to that point.  That might be the most reasonable API now that I think about it.

**Comment by abarth** [comment]
> So should the tap detector invoke the tap callback but not declare victory?

I think desired behavior is that the possibility of double-tap delays the tap callback until there can no longer be a double-tap.

> If we have a tap detector in place but not a double tap detector, and someone does a double tap, should two separate tap event callbacks be invoked? 

Yes.  Consider a button that increments a counter.  You should be able to spam the button with taps and the counter should keep incrementing.

> How about the same scenario but you add a double tap detector?

In that case, you should get a double-tap but no tap.

@Hixie

**Comment by krisgiesing** [comment]
I think people would be really confused and annoyed if their tap gestures accrued latency due to double tap detectors elsewhere, but that may be me remembering all the headaches with tap latency on mobile HTML5.

I think just passing on the number of consecutive taps in the tap callback gives you all the flexibility you'd need with a simple API and no latency overhead.  Conflicts between single and double tap recognition then become semantic conflicts the app has to resolve anyway.

**Comment by Hixie** [comment]
@krisgiesing The latency only comes in if you're in an arena with a doubletap detector (which is only when doubletap does something).

Tap vs Scroll being different that Tap vs Double Tap is weird though...

**Comment by krisgiesing** [comment]
Re-doing this on a different branch.

**Pull request closed without merging**